lib/charms/operator_libs_linux/v0/sysctl.py

```diff
@@ -0,0 +1,285 @@
+"""
+
+import logging
+import re
+from pathlib import Path
+from subprocess import STDOUT, CalledProcessError, check_output
+from typing import Dict, List
+
+logger = logging.getLogger(__name__)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "17a6cd4d80104d15b10f9c2420ab3266"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 2
+
+CHARM_FILENAME_PREFIX = "90-juju-"
+SYSCTL_DIRECTORY = Path("/etc/sysctl.d")
+SYSCTL_FILENAME = Path(f"{SYSCTL_DIRECTORY}/95-juju-sysctl.conf")
+SYSCTL_HEADER = f"""# This config file was produced by sysctl lib v{LIBAPI}.{LIBPATCH}
+#
+# This file represents the output of the sysctl lib, which can combine multiple
+# configurations into a single file like.
+"""
+
+
+class Error(Exception):
+    """Base class of most errors raised by this library."""
+
+    def __repr__(self):
+        """Represent the Error."""
+        return "<{}.{} {}>".format(type(self).__module__, type(self).__name__, self.args)
+
+    @property
+    def name(self):
+        """Return a string representation of the model plus class."""
+        return "<{}.{}>".format(type(self).__module__, type(self).__name__)
+
+    @property
+    def message(self):
+        """Return the message passed as an argument."""
+        return self.args[0]
+
+
+class SysctlError(Error):
+    """Raised when there's an error running sysctl command."""
+
+
+class SysctlPermissionError(Error):
+    """Raised when there's an error applying values in sysctl."""
+
+
+class ValidationError(Error):
+    """Exception representing value validation error."""
+
+
+class Config(Dict):
+    """Represents the state of the config that a charm wants to enforce."""
+
+    def __init__(self, name: str) -> None:
+        self.name = name
+        self._data = self._load_data()
+
+    def __contains__(self, key: str) -> bool:
+        """Check if key is in config."""
+        return key in self._data
+
+    def __len__(self):
+        """Get size of config."""
+        return len(self._data)
+
+    def __iter__(self):
+        """Iterate over config."""
+        return iter(self._data)
+
+    def __getitem__(self, key: str) -> int:
+        """Get value for key form config."""
+        return self._data[key]
+
+    @property
+    def charm_filepath(self) -> Path:
+        """Name for resulting charm config file."""
+        return SYSCTL_DIRECTORY / f"{CHARM_FILENAME_PREFIX}{self.name}"
+
+    def update(self, config: Dict[str, dict]) -> None:
+        """Update sysctl config options with a desired set of config params.
+
+        Args:
+            config: dictionary with keys to update:
+        ```
+        {"vm.swappiness": {"value": 10}, ...}
+        ```
+        """
+        self._parse_config(config)
+
+        # NOTE: case where own charm calls update() more than once. Remove first so
```

Comment:
With calling just `_merge(own_fille)` and reloading keys from `95-juju-sysctl.conf` a charm can update successfully the config. If something fails, the previously applied file from charm will still be there.

Reply:
Yes, but now I don't remove. In that scenario, the last call of update will run successfully, which means that the file will be created and merged. If other charm comes along, it will get a validation error.